File: src/cmcandy/solutions/_0055.java
package cmcandy.solutions;

/**
 * Created by CHEN on 2020/5/14.
 */
public class _0055 {

    static class Solution {
        public boolean canJump(int[] nums) {

            int len = nums.length;
            boolean tmp = true;
            //搜索数组中有没有0
            for(int i =0;i<len-1;i++){
                if(nums[i]==0){
                    tmp =false;
                    for(int j =i-1;j>=0;j--){
                        if(nums[j]>i-j) tmp=true;
                    }
                    if (tmp==false) return false;
                }
            }
            return true;
        }
    }


    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.canJump(new int[]{3,2,1,0,4}) );
    }
}

File: src/cmcandy/C_Language_Answers/_0055.cpp
#include <iostream>
#include <vector>
#include <limits.h>
using namespace ::std;
class Solution
{
public:
    bool canJump(vector<int> &nums)
    {
        int len = nums.size();
        if (len == 1)
            return true;
        int step = nums[0];
        //最后一个不需要判断，能走到即是胜利
        for (int i = 0; i < len - 1; i++)
        {
            //如果比step大说明按这格可以跳得更远
            if (nums[i] > step)
                step = nums[i];
            //判断能不能到跳过这个0
            if (step == 0 && nums[i] == 0)
                return false;
            step--;
        }
        return true;
    }
};